### DISCUSSION TITLE
Cloudflare Workers KV in /api routes

### DISCUSSION THREAD

**Discussion by marshallswain** [nuxt/nuxt#16123] [Questions] [answered]
I'm looking for a way to access Workers KV and other Cloudflare bindings inside /api routes during dev.  Is there a way to use Miniflare for the dev server?

The new `modules` format for Cloudflare Workers has the below format, with an `env` object as the second argument.  The `env` object contains all bindings for the request, including KV store access and environment variables.  Is there a way to create some kind of transport wrapper for the dev server?  it would also need to be compatible with the deployed h3 server, I suppose.  Currently, I work around this issue by deploying my Nuxt 3 site to a separate Cloudflare account.  Cloudflare Workers do not allow making HTTP requests between workers on the same account as a security measure.  Instead, they support worker bindings, which basically allow access to another worker as an environment variable inside the `env` object.  Access to a bound worker would happen like the below example.  Requests across worker bindings happen internally and don't make actual HTTP requests.

```ts
export default {
  fetch(request, env) {
    /* bindings found in `env` */
    env.MY_OTHER_WORKER.fetch(...)
  }
}

```

I realize that it's possible to use Miniflare in `nuxi preview`, but that makes it really difficult to write Workers code efficiently.  Any pointers along these lines would be great.  If it's not possible, my two-account workaround works well, for now.  I'm looking forward to making the process more streamlined and efficient.

**Comment by marshallswain** [comment, marked as answer]
# Work in Progress

Update, now that [Miniflare 3 is available and built on workersd](https://blog.cloudflare.com/miniflare-and-workerd/), this needs to be re-evaluated. https://blog.cloudflare.com/miniflare-and-workerd/

After reviewing the source of [nuxt/framework](https://github.com/nuxt/framework), [unjs/nitro](https://github.com/unjs/nitro), and [unjs/h3](https://github.com/unjs/h3), I've confirmed that when using Cloudflare's new `module` worker format it's not currently possible to use Workers KV in Nuxt components or api routes.  (Workers KV does work with the current deployment preset for Cloudflare, since it still targets the `worker` format.)  Even if it were possible to create development preset adapters for Nuxt (similar to deployment presets), the MiniFlare API requires injecting a worker script as a string.  There's no way to use it as a wrapper around nitro/h3.

I'm going to use this discussion to document possibilities for adding support for Cloudflare APIs into Nuxt.  These are the issues that need to be addressed:

- Use Cloudflare APIs in dev
- Access Cloudflare-specific `env` variables with `useRuntimeConfig()` in production
- Compatibility with the Cloudflare Workers `module` format

## Use Cloudflare APIs in Dev Mode

As already discussed, above, Miniflare cannot be directly used for Cloudflare development.  Instead, a mock adapter will need to be created for each desired API.  There are two interesting types of APIs for Cloudflare Worker interaction in Nuxt:

1. **Service Bindings**: allowing direct communication with other workers on the same Cloudflare account.
2. **Cache-related Bindings**: which provide access to Workers KV and Durable Objects.

In a modern Cloudflare worker (read: esm format), the bindings are all found in the `env` object which is passed as the second argument on every incoming `fetch` request.  Since all environment variables are stored there, it makes sense to augment the Nuxt Runtime Config with the bindings.

### 1. Mocking Service Bindings

Service bindings allow calling `fetch` on the provided binding variable, as shown here:

```js
export default {
  fetch(request, env) {
    /* bindings found in `env` */
    env.MY_OTHER_WORKER.fetch(...)
  }
}
```

These bindings are pre-configured in the `wrangler.toml` file.  An example binding looks like the example, below, which is taken at the end of [this page in the Wrangler configuration docs](https://developers.cloudflare.com/workers/wrangler/configuration/)

```toml
# A list of other Cloudflare services bound to this service.
# @default `[]`
# @param {string} binding The binding name used to refer to the bound service.
# @param {string} service The name of the service.
# @param {string} environment The environment of the service (For example, production, staging, etc) (optional).
services = [
  { binding = "TEST_BINDING", service = "", environment = "" }
]
```

So the ideal utility would be a plugin that reads the services from the `wrangler.toml` file and automatically sets up `$fetch` wrappers that implement the above API.  The only missing ingredient is URL configuration.

One possible solution to configuring the URL would be to use an environment variable convention. So, if your binding name is `TEST_BINDING`, you can create a `TEST_BINDING_URL` variable in the `runtimeConfig` to be read by the plugin.  The downside to this is that it separates the implementation from the plugin, requiring multiple places to be updated.

An alternative would be to provide a configuration object to the Nuxt plugin.  Something like

```js
{ 
  services: {
    TEST_BINDING: 'http://localhost:3334'
  }
}
```

I prefer the plugin/configuration object as a solution.  This plugin will only be used for development, so there's no need to clutter the `runtimeConfig`.

> Aside: the same pattern can probably be used for local worker development, since multi-worker development is currently not supported by Wrangler.  More information can be found in [this issue](https://github.com/cloudflare/wrangler2/issues/1182).

### 2. Mocking Workers KV and Durable Objects

Similar to the solution for Service Bindings, above, the runtimeConfig will need to be augmented with bindings found in the `wrangler.toml` file.  There's also a utility for each type of cache found in the Miniflare repo.

#### Workers KV

Workers KV configuration:

```toml
# These specify any Workers KV Namespaces you want to
# access from inside your Worker.
# To learn more about KV namespaces, refer to:
# https://developers.cloudflare.com/workers/learning/how-kv-works
# @default `[]`
# @param {string} binding The binding name used to refer to the KV namespace
# @param {string} id The ID of the KV namespace at the edge
# @param {string} preview_id The ID of the KV namespace used during `wrangler dev`
# not inherited
kv_namespaces = [
  { binding = "TEST_NAMESPACE", id = "", preview_id = "" }
]
```

Miniflare's Worker's KV module [is found here](https://github.com/cloudflare/miniflare/tree/master/packages/kv).

#### Durable Objects

Durable Objects Configuration

```toml
# A list of Durable Objects that your Worker should be bound to.
# To learn more about Durable Objects, refer to:
# https://developers.cloudflare.com/workers/learning/using-durable-objects
# @default `{bindings:[]}`
# @param {string} name The name of the binding used to refer to the Durable Object
# @param {string} class_name The exported class name of the Durable Object
# @param {string} script_name The script where the Durable Object is defined (if it is external to this Worker)
# not inherited
[durable_objects]
  bindings = [
    { name = "TEST_OBJECT", class_name = "", script_name = "" }
  ]
```

Durable Object Miniflare utility [is found here](https://github.com/cloudflare/miniflare/tree/master/packages/durable-objects).  I notice that the API differs from actual Durable Objects usage, so it might need to be adapted to use stubs.  Needs verification.


## Update Runtime Config in Production

A solution for this is described in the next section.  Basically, we need to create a new deployment preset for `cloudflare-esm` which also provides the `env` and `ctx` objects in the request.

## Support the Cloudflare Workers `module` Format

A new `cloudflare-esm` deployment preset will be needed. to house the API changes.  I've started an example runtime entry for the new preset, below the summary of breaking changes.

### Format Differences: Service Worker vs new ESM Format

With Wrangler 2.0, the format of a Cloudflare Worker has changed.  Here's an example of the old `worker` format.  Note the following, because they all change with the `module` format:

- `addEventListener` registers a handler, which receives an `event` object.  (The event object contains the initial `request` instance with some other helpers)
- Environment variables are bound as globals.
- `event.respondWith(new Response(...))` sends a response:

```js
// Old `worker` Format
addEventListener('fetch', event => {
  const data = { hello: 'world' };
  const json = JSON.stringify(data, null, 2);

  // Access globally-bound environment variable.
  console.log(MY_ENV_VAR)

  return event.respondWith(
    new Response(json, {
      headers: {
        'content-type': 'application/json;charset=UTF-8',
      },
    })
  );
});
```

The new `module` format is simpler.  Note the following changes:

- The handler is now a `fetch` method in an object.  Instead of `event`, it receives the `request`, `env`, and `ctx` objects.
- Environment variables are in the `env` object.  There is no more global access.
- The Response instance is directly returned.  `event.respondWith` is gone:

```js
// New `module` Format
export default {
  fetch(request, env, ctx) {
    const data = { hello: 'world' };
    const json = JSON.stringify(data, null, 2);

    // Access environment variables on `env`
    console.log(env.MY_ENV_VAR)

    return new Response(json, {
      headers: {
        'content-type': 'application/json;charset=UTF-8',
      },
    })
  }
}
```

### Runtime Preset for `cloudflare-esm`

I started adapting the [existing `cloudflare` preset](https://github.com/unjs/nitro/blob/main/src/runtime/entries/cloudflare.ts) to check for potential blockers.  The @cloudflare/kv-asset-handler package has a [new API for the esm format](https://github.com/cloudflare/kv-asset-handler#es-modules).    Here is an adaptation of the Cloudflare preset (which I haven't tested, yet, since the service-worker build seems to be broken in the latest Nuxt release candidate, right now.)

```ts
import '#internal/nitro/virtual/polyfill'
import { getAssetFromKV, mapRequestToAsset } from '@cloudflare/kv-asset-handler'
import { withoutBase } from 'ufo'
import { requestHasBody, useRequestBody } from '../utils'
import { nitroApp } from '../app'
import { useRuntimeConfig } from '#internal/nitro'

// Bundled by Wrangler 2. See https://github.com/cloudflare/kv-asset-handler#asset_manifest-required-for-es-modules
import manifestJSON from '__STATIC_CONTENT_MANIFEST'
const assetManifest = JSON.parse(manifestJSON)

export default {
  fetch(request, env, ctx) {
    try {
      // TODO: Update to new API: https://github.com/cloudflare/kv-asset-handler#es-modules
      return await getAssetFromKV({
        request,
        waitUntil(promise) {
          return ctx.waitUntil(promise)
        }
      }, {
        cacheControl: assetsCacheControl,
        mapRequestToAsset: baseURLModifier,
        ASSET_NAMESPACE: env.__STATIC_CONTENT,
        ASSET_MANIFEST: assetManifest,
      })
    } catch (_err) {
      // Ignore
    }

    const url = new URL(request.url)
    let body
    if (requestHasBody(request)) {
      body = await useRequestBody(request)
    }

    const r = await nitroApp.localCall({
      event: { request },
      url: url.pathname + url.search,
      host: url.hostname,
      protocol: url.protocol,
      headers: Object.fromEntries(request.headers.entries()),
      method: request.method,
      redirect: request.redirect,
      body
    })

    return new Response(r.body, {
      headers: normalizeOutgoingHeaders(r.headers),
      status: r.status,
      statusText: r.statusText
    })
  }
}

function assetsCacheControl(_request) {
  // TODO: Detect public asset bases
  // if (request.url.startsWith(buildAssetsURL())) {
  //   return {
  //     browserTTL: 31536000,
  //     edgeTTL: 31536000
  //   }
  // }
  return {}
}

const baseURLModifier = (request: Request) => {
  const url = withoutBase(request.url, useRuntimeConfig().app.baseURL)
  return mapRequestToAsset(new Request(url, request))
}

function normalizeOutgoingHeaders(headers: Record<string, string | string[] | undefined>) {
  return Object.entries(headers).map(([k, v]) => [k, Array.isArray(v) ? v.join(',') : v])
}
```

**Reply by atinux** [reply]
Wow thank you for your answer, could you open an issue or even PR in nitro to move the conversation going into an improvement into Nuxt. 😊

**Comment by paulmelero** [comment]
if not in `dev`, did you find a way, in general, to access other bindings inside `server/api`?

I cannot find a way of accessing other existing workers from my account with the current Nuxt APIs.

**Reply by paulmelero** [reply]
Basically, I was trying to deploy my Nuxt app to Cloudflare workers and I also wanted to use other workers as external HTTP services. I made it work using a regular HTTP(s) call in local, tested using `miniflare`, but then the HTTP call does not work when deployed. A worker making an HTTP call to another worker is forbidden, as far as I know... 🤷🏽‍♂️

**Reply by marshallswain** [reply]
Correct, making plain HTTP requests between workers on the same account is blocked by Cloudflare, intentionally.  My huge comment, above, outlines how to make all this stuff work.  I'm going to start working on it this weekend.

**Reply by paulmelero** [reply]
It would be lovely if you make it work. I tried to edit the `nitro` preset but couldn't make it work on my side 🙏🏽

**Comment by timhanlon** [comment]
Thanks so much for your investigation @marshallswain, helped me out a lot.

Here's a snippet I wrote yesterday for a project of mine which works in dev, Miniflare and a Nuxt app deployed to Cloudflare Workers.

```
const getKVNamespace = async (ns) => {
  if (process.dev) {
    const { KVNamespace } = await import('@miniflare/kv')
    const { FileStorage } = await import('@miniflare/storage-file')
    return new KVNamespace(new FileStorage(`./.mf/kv/${ns}`))
  } else {
    return self[ns]
  }
}

export { getKVNamespace }
```

**Reply by marshallswain** [reply]
Thank you, @timhanlon.  I intend to both start and finish this work up by Christmas. I'm sure your snippet will come in handy.

**Comment by shaharyarrahmad** [comment]
@marshallswain any update on this?

**Reply by marshallswain** [reply]
Support will be added when these PRs are merged: 

- https://github.com/unjs/nitro/pull/681
- https://github.com/unjs/nitro/pull/677

**Reply by gmesml** [reply]
A TODO repo using nuxt3 would come handy.

**Reply by gmesml** [reply]
Looks like bindings can be accessed now on env or ctx param: https://github.com/nuxt/nuxt/issues/18599#issuecomment-1468524564
With the release of 3.3
